ntoskrnl/inbv/inbv.c

```diff
@@ -932,18 +953,145 @@ DisplayBootBitmap(IN BOOLEAN TextMode)
+        InbvRotBarInit();
     }
 #endif
 }
 
+#ifdef INBV_ROTBAR_IMPLEMENTED
+static
+VOID
+NTAPI
+InbvRotationThread(
+    _In_ PVOID Context)
+{
+    ULONG X, Y, Index = 0, Total = 18;
+
+    InbvAcquireLock();
+    X = ProgressBarLeft + 2;
+    Y = ProgressBarTop + 2;
+    InbvReleaseLock();
+
+    while (TRUE)
+    {
+        InbvAcquireLock();
+
+        if (InbvDisplayState != INBV_DISPLAY_STATE_OWNED || RotBarSelection != RB_UNSPECIFIED)
+        {
+            InbvReleaseLock();
+            break;
+        }
+
+        if (Index >= 3)
+        {
+            /* Fill previous bar position */
+            VidSolidColorFill(X + ((Index - 3) * 8), Y, (X + ((Index - 3) * 8)) + 8 - 1, Y + 9 - 1, 0);
+        }
+        if (Index < Total - 1)
+        {
+            /* Draw the progress bar bit */
+            if (Index < 2)
+            {
+                /* Appearing from the left */
+                VidBufferToScreenBlt(RotBarBuffer + 8 * (2 - Index), X, Y, 22 - 8 * (2 - Index), 9, 24);
+            }
+            else if (Index >= Total - 3)
+            {
+                /* Hiding to the right */
+                VidBufferToScreenBlt(RotBarBuffer, X + ((Index - 2) * 8), Y, 22 - 8 * (4 - (Total - Index)), 9, 24);
+            }
+            else
+            {
+                VidBufferToScreenBlt(RotBarBuffer, X + ((Index - 2) * 8), Y, 22, 9, 24);
+            }
+        }
+
+        InbvReleaseLock();
+
+        /* Wait for a bit */
+        KeStallExecutionProcessor(ROT_BAR_TIME);
+
+        Index = (Index + 1) % Total;
+    }
+}
+
+VOID
+NTAPI
+INIT_FUNCTION
+InbvRotBarInit(VOID)
+{
+    NTSTATUS Status;
+    HANDLE ThreadHandle;
+
+    if (!InbvBootDriverInstalled || RotBarSelection == RB_UNSPECIFIED)
+        return;
+
+    RotBarThread = NULL;
+
+    ThreadHandle = INVALID_HANDLE_VALUE;
+    Status = PsCreateSystemThread(&ThreadHandle,
+                                  SYNCHRONIZE,
+                                  NULL,
+                                  NULL,
+                                  NULL,
+                                  InbvRotationThread,
+                                  NULL);
+
+    if (NT_SUCCESS(Status) && ThreadHandle != NULL && ThreadHandle != INVALID_HANDLE_VALUE)
+    {
+        Status = ObReferenceObjectByHandle(ThreadHandle,
+                                           SYNCHRONIZE,
+                                           PsThreadType,
+                                           KernelMode,
+                                           &RotBarThread,
+                                           NULL);
+        if (NT_SUCCESS(Status) && RotBarThread)
+            RotBarThreadActive = TRUE;
+        ObCloseHandle(ThreadHandle, KernelMode);
+    }
+}
+
+VOID
+NTAPI
+INIT_FUNCTION
+InbvRotBarStop(VOID)
+{
+    NTSTATUS Status;
+
+    if (!RotBarThread)
+    {
+        RotBarThreadActive = FALSE;
+        return;
+    }
+
+    Status = KeWaitForSingleObject(RotBarThread,
```

Comment:
Before this call (but after the previous test above) you would set the flag (I've talked about in the thread function) to FALSE, then wait for the thread to finish as you seem to do here.

Reply:
Ok, moved variable assignment here.